Add unit tests for Tseitin CNF conversion and implement missing `equals` method for `Sym`.

gbasler opened the pull request:
As promised in #4078.

The tests already paid off: figured out that I forgot to add `equals` (shame on me)...

Review by @adriaanm

adriaanm commented:
Thanks! Could you split up the commit into the bugfix for the earlier PR (equals method) and the test? Why is the solver code duplicated in the test? Can we refactor the actual solver to make it testable directly?

gbasler committed:
Bugfix: Implement missing `equals` method for `Sym`.

gbasler committed:
Add unit tests for Tseitin CNF conversion.

We compare the results using the Tseitin transformation with the results of a
conversion via expansion of the formula (using distributive laws), e.g.,

```
             +-------+
             |Formula|
             +---+---+
                 |
     +-----------+-----------+
     |                       |
     v                       v
+---------+              +-------+
|Expansion|              |Tseitin|
+----+----+              +---+---+
     |        +-----+        |
     +------->| =?= |<-------+
              +-----+
```

both methods should deliver the same results (i.e., models).

gbasler commented:
I've split the commit into two parts.

The current solver implementation is coupled with it's usage, I used a suggestion by @retronym to decouple the code. I don't see a better way right now since the traits are otherwise not accessible.
Actually there should not be any duplicated code, the two functions that I added are not needed anymore by the solver that is used in the pattern matcher.

adriaanm commented:
LGTM, thanks for splitting them up. I'm going to merge this so we can get the improvement into 2.11.5
